src/main.rs

```diff
@@ -298,52 +299,62 @@ fn decode_and_print_defmt_logs(
     stream_decoder: &mut dyn StreamDecoder,
     locations: Option<&Locations>,
     current_dir: &Path,
-    opts: &cli::Opts,
+    shorten_paths: bool,
+    encoding_can_recover: Option<bool>,
 ) -> anyhow::Result<()> {
     loop {
         match stream_decoder.decode() {
-            Ok(frame) => {
-                // NOTE(`[]` indexing) all indices in `table` have already been verified to exist in
-                // the `locations` map
-                let (file, line, mod_path) = locations
-                    .map(|locations| &locations[&frame.index()])
-                    .map(|location| {
-                        let path = if let Ok(relpath) = location.file.strip_prefix(&current_dir) {
-                            relpath.display().to_string()
-                        } else {
-                            let dep_path = dep::Path::from_std_path(&location.file);
-
-                            if opts.shorten_paths {
-                                dep_path.format_short()
-                            } else {
-                                dep_path.format_highlight()
-                            }
-                        };
-
-                        (
-                            Some(path),
-                            Some(location.line as u32),
-                            Some(&*location.module),
-                        )
-                    })
-                    .unwrap_or((None, None, None));
-
-                // Forward the defmt frame to our logger.
-                defmt_decoder::log::log_defmt(&frame, file.as_deref(), line, mod_path);
-            }
-
-            Err(defmt_decoder::DecodeError::UnexpectedEof) => break,
-
-            Err(defmt_decoder::DecodeError::Malformed) => {
-                log::error!("failed to decode defmt data");
-                return Err(defmt_decoder::DecodeError::Malformed.into());
-            }
+            Ok(frame) => forward_to_logger(&frame, locations, current_dir, shorten_paths),
+            Err(DecodeError::UnexpectedEof) => break,
+            Err(DecodeError::Malformed) => match encoding_can_recover {
+                // if recovery is impossible, or we don't know if it is, abort
+                Some(false) | None => return Err(DecodeError::Malformed.into()),
```

Comment:
observation: the `None` case is unreachable because a `StreamDecoder` can only be constructed when an `ElfTable` exists and this encoding information can always be extracted from `ElfTable`

Reply:
I would suggest extracting the encoding info from [this line](https://github.com/knurling-rs/probe-run/blob/f03dcf195ca538dc9ec4afe0b6938e3017fcb181/src/main.rs#L226) so you would end up with a option of a tuple that contains the stream decoder and the encoding info.

Reply:
Kk. What about now?